src/lightning_lite/utilities/apply_func.py

```diff
@@ -0,0 +1,339 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities used for collections."""
+
+import dataclasses
+from abc import ABC
+from collections import defaultdict, OrderedDict
+from copy import deepcopy
+from functools import partial
+from typing import Any, Callable, List, Mapping, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import torch
+from torch import Tensor
+
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+_BLOCKING_DEVICE_TYPES = ("cpu", "mps")
+
+
+def to_dtype_tensor(
+    value: Union[int, float, List[Union[int, float]]], dtype: torch.dtype, device: Union[str, torch.device]
+) -> Tensor:
+    return torch.tensor(value, dtype=dtype, device=device)
+
+
+def from_numpy(value: np.ndarray, device: Union[str, torch.device]) -> Tensor:
+    return torch.from_numpy(value).to(device)
+
+
+CONVERSION_DTYPES: List[Tuple[Any, Callable[[Any, Any], Tensor]]] = [
+    # bool -> uint8 as bool -> torch.bool triggers RuntimeError: Unsupported data type for NCCL process group
+    (bool, partial(to_dtype_tensor, dtype=torch.uint8)),
+    (int, partial(to_dtype_tensor, dtype=torch.int)),
+    (float, partial(to_dtype_tensor, dtype=torch.float)),
+    (np.ndarray, from_numpy),
+]
+
+
+def _is_namedtuple(obj: object) -> bool:
+    # https://github.com/pytorch/pytorch/blob/v1.8.1/torch/nn/parallel/scatter_gather.py#L4-L8
+    return isinstance(obj, tuple) and hasattr(obj, "_asdict") and hasattr(obj, "_fields")
+
+
+def _is_dataclass_instance(obj: object) -> bool:
+    # https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions
+    return dataclasses.is_dataclass(obj) and not isinstance(obj, type)
+
+
+def apply_to_collection(
```

Comment:
Opened https://github.com/Lightning-AI/utilities/pull/32 for this

We just need to release it to PyPI

Reply:
Can it be done in parallel? I don't think it needs to block this. There will be a CI PR needed to add this as dependency and so on, which can take some time.